File: packages/zerobox/src/flags.ts
import type { SandboxOptions } from "./types.js";

/** Build CLI flags from SandboxOptions. */
export function buildFlags(options: SandboxOptions): string[] {
  const flags: string[] = [];

  // Collect secret hosts for network permission merging.
  const secretHosts: string[] = [];

  if (options.secrets) {
    for (const [key, config] of Object.entries(options.secrets)) {
      flags.push("--secret", `${key}=${config.value}`);
      if (config.hosts.length > 0) {
        flags.push("--secret-host", `${key}=${config.hosts.join(",")}`);
        secretHosts.push(...config.hosts);
      }
    }
  }

  if (options.strictSandbox) {
    flags.push("--strict-sandbox");
  }
  if (options.debug) {
    flags.push("--debug");
  }

  if (options.allowAll) {
    flags.push("--allow-all");
  } else if (options.noSandbox) {
    flags.push("--no-sandbox");
  } else {
    const explicit = Array.isArray(options.profile)
      ? options.profile
      : options.profile
        ? [options.profile]
        : [];
    const profiles = explicit.length > 0 ? explicit : ["workspace"];
    for (const p of profiles) {
      flags.push("--profile", p);
    }
    if (options.allowRead?.length) {
      flags.push(`--allow-read=${options.allowRead.join(",")}`);
    }
    if (options.denyRead?.length) {
      flags.push(`--deny-read=${options.denyRead.join(",")}`);
    }
    if (options.allowWrite?.length) {
      flags.push(`--allow-write=${options.allowWrite.join(",")}`);
    }
    if (options.denyWrite?.length) {
      flags.push(`--deny-write=${options.denyWrite.join(",")}`);
    }

    // Merge secret hosts into allowNet (secrets auto-enable network for their hosts).
    // The CLI also does this, but we emit it here so --allow-net reflects the full picture.
    let effectiveAllowNet = options.allowNet;
    if (secretHosts.length > 0) {
      if (effectiveAllowNet === true) {
        // Already allowing all network.
      } else if (Array.isArray(effectiveAllowNet)) {
        effectiveAllowNet = [...effectiveAllowNet, ...secretHosts];
      } else {
        // Network was not explicitly enabled; secrets handle it via --secret-host.
        // Don't emit --allow-net here — the CLI enables network implicitly for secret hosts.
      }
    }

    if (effectiveAllowNet === true) {
      flags.push("--allow-net");
    } else if (Array.isArray(effectiveAllowNet) && effectiveAllowNet.length > 0) {
      flags.push(`--allow-net=${effectiveAllowNet.join(",")}`);
    }
    if (options.denyNet?.length) {
      flags.push(`--deny-net=${options.denyNet.join(",")}`);
    }
  }

  // Env flags — emitted for all modes (including allowAll).
  if (options.allowEnv === true) {
    flags.push("--allow-env");
  } else if (Array.isArray(options.allowEnv) && options.allowEnv.length > 0) {
    flags.push(`--allow-env=${options.allowEnv.join(",")}`);
  }
  if (options.denyEnv?.length) {
    flags.push(`--deny-env=${options.denyEnv.join(",")}`);
  }
  if (options.env) {
    for (const [key, value] of Object.entries(options.env)) {
      flags.push("--env", `${key}=${value}`);
    }
  }

  if (options.bindMounts?.length) {
    for (const m of options.bindMounts) {
      const spec = m.readOnly ? `${m.host}:${m.sandbox}:ro` : `${m.host}:${m.sandbox}`;
      flags.push("--bind-mount", spec);
    }
  }

  if (options.cwd) {
    flags.push("-C", options.cwd);
  }

  if (options.restore) {
    flags.push("--restore");
  } else if (options.snapshot) {
    flags.push("--snapshot");
  }
  if (options.snapshotPaths?.length) {
    flags.push(`--snapshot-path=${options.snapshotPaths.join(",")}`);
  }
  if (options.snapshotExclude?.length) {
    flags.push(`--snapshot-exclude=${options.snapshotExclude.join(",")}`);
  }

  return flags;
}

File: sdks/python/src/zerobox/flags.py
from __future__ import annotations

from typing import Any, Union

from .options import SandboxOptions, normalize_options


def _add_csv(flags: list[str], name: str, values: Union[list[str], None]) -> None:
    if values:
        flags.append(f"--{name}={','.join(values)}")


def _add_allow(flags: list[str], name: str, value: Union[bool, list[str], None]) -> None:
    if value is True:
        flags.append(f"--{name}")
    elif isinstance(value, list):
        _add_csv(flags, name, value)


def _profile_list(profile: Union[str, list[str], None]) -> list[str]:
    if isinstance(profile, list) and profile:
        return list(profile)
    if isinstance(profile, str) and profile:
        return [profile]
    return ["workspace"]


def build_flags(options: Union[SandboxOptions, dict[str, Any], None]) -> list[str]:
    """Convert SandboxOptions → CLI flags. Byte-compatible with TS `buildFlags`."""
    o = normalize_options(options)
    flags: list[str] = []
    secret_hosts: list[str] = []

    if o.secrets:
        for key, cfg in o.secrets.items():
            flags.extend(["--secret", f"{key}={cfg.value}"])
            if cfg.hosts:
                flags.extend(["--secret-host", f"{key}={','.join(cfg.hosts)}"])
                secret_hosts.extend(cfg.hosts)

    if o.strict_sandbox:
        flags.append("--strict-sandbox")
    if o.debug:
        flags.append("--debug")

    if o.allow_all:
        flags.append("--allow-all")
    elif o.no_sandbox:
        flags.append("--no-sandbox")
    else:
        for p in _profile_list(o.profile):
            flags.extend(["--profile", p])

        _add_csv(flags, "allow-read", o.allow_read)
        _add_csv(flags, "deny-read", o.deny_read)
        _add_csv(flags, "allow-write", o.allow_write)
        _add_csv(flags, "deny-write", o.deny_write)

        # Secret hosts implicitly allow network for their domains. When
        # allow_net is already a list, append the hosts. When True, leave
        # alone. When unset, let the CLI gate network via secret-host.
        allow_net = o.allow_net
        if secret_hosts and isinstance(allow_net, list):
            allow_net = [*allow_net, *secret_hosts]

        _add_allow(flags, "allow-net", allow_net)
        _add_csv(flags, "deny-net", o.deny_net)

    _add_allow(flags, "allow-env", o.allow_env)
    _add_csv(flags, "deny-env", o.deny_env)
    if o.env:
        for key, value in o.env.items():
            flags.extend(["--env", f"{key}={value}"])

    if o.cwd:
        flags.extend(["-C", o.cwd])

    if o.restore:
        flags.append("--restore")
    elif o.snapshot:
        flags.append("--snapshot")
    _add_csv(flags, "snapshot-path", o.snapshot_paths)
    _add_csv(flags, "snapshot-exclude", o.snapshot_exclude)

    return flags
